jsonpointer.py

```diff
@@ -29,40 +29,50 @@
 # (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 # THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #
+from __future__ import annotations
 
 """ Identify specific nodes in a JSON document (RFC 6901) """
 
-from __future__ import unicode_literals
-
 # Will be parsed by setup.py to determine package metadata
-__author__ = 'Stefan Kögl <stefan@skoegl.net>'
-__version__ = '2.3'
-__website__ = 'https://github.com/stefankoegl/python-json-pointer'
-__license__ = 'Modified BSD License'
+__author__ = "Stefan Kögl <stefan@skoegl.net>"
+__version__ = "3.0"
+__website__ = "https://github.com/stefankoegl/python-json-pointer"
+__license__ = "Modified BSD License"
 
 
-try:
-    from itertools import izip
-    str = unicode
-    encode_str = lambda u: u.encode("raw_unicode_escape")
-except ImportError:  # Python 3
-    izip = zip
-    encode_str = lambda u: u
+import copy
+import re
+from collections.abc import Mapping, Sequence
+from itertools import tee, chain
+from typing import *
 
-try:
-    from collections.abc import Mapping, Sequence
-except ImportError:  # Python 3
-    from collections import Mapping, Sequence
+T = TypeVar("T")
+TJsonPointer = TypeVar("TJsonPointer", bound="JsonPointer")
```

Comment:
Used in places like `@classmethod`, where `def foo(cls: Type[TJsonPointer]) -> TJsonPointer` is dependent on the subclass. It's a small detail, but typing it correctly is really nice to people downstream.

In 3.11, PEP 673 makes this even easier, but since this was supporting every python version not EOL I didn't leverage it https://peps.python.org/pep-0673/